File: crates/suspect-codegen/src/php_sdk/Validation.php
<?php
declare(strict_types=1);

namespace __NAMESPACE__;

require_once __DIR__ . '/ValidationV2.php';
require_once __DIR__ . '/ValidationResources.php';

/** Completed invalidity and incomplete evaluation are distinct exception kinds. */
class ValidationError extends \RuntimeException
{
    public function __construct(
        public readonly string $kind,
        public readonly string $source,
        public readonly string $instancePath,
        string $message,
    ) { parent::__construct($kind . ' at ' . JsonError::display($source) . ' ' . JsonError::display($instancePath) . ': ' . JsonError::display($message)); }
}

/** Parse immutable program literals on first use, preserving unvisited-operand semantics. @internal */
final class ValidationLiteral
{
    private ?JsonValue $parsed = null;
    public function __construct(private readonly string $json) {}
    public function value(?CallControl $control): JsonValue
    {
        return $this->parsed ??= JsonValue::parse($this->json, new JsonLimits(control: $control));
    }
}

/** Immutable lowered instruction; only the generator constructs this metadata. @internal */
final readonly class ValidationInstruction
{
    /**
     * @param list<string> $types
     * @param list<int> $targets
     * @param list<array{string, int}> $properties
     * @param list<string> $names
     * @param list<string> $declared
     * @param list<ValidationLiteral> $literals
     * @param list<array{string,list<string>}> $requirements
     * @param list<array{string,int}> $dependencies
     * @param list<PatternProperty> $patterns
     */
    public function __construct(
        public string $op,
        public string $source,
        public bool $truth = false,
        public array $types = [],
        public int $target = -1,
        public array $targets = [],
        public array $properties = [],
        public array $names = [],
        public array $declared = [],
        public int $start = 0,
        public ?JsonNumber $number = null,
        public bool $maximum = false,
        public bool $exclusive = false,
        public string $countTarget = '',
        public ?ValidationLiteral $constant = null,
        public array $literals = [],
        public ?PatternProgram $pattern = null,
        public int $condition = -1,
        public ?int $thenTarget = null,
        public ?int $elseTarget = null,
        public array $requirements = [],
        public array $dependencies = [],
        public ?JsonNumber $containsMinimum = null,
        public ?JsonNumber $containsMaximum = null,
        public array $patterns = [],
        public int $initialResource = -1,
        public ?string $anchor = null,
    ) {}
}

/** @internal */
final readonly class ValidationNode
{
    /** @param list<ValidationInstruction> $checks */
    public function __construct(public string $source, public array $checks) {}
}

/** Portable ECMA-262 Unicode Thompson NFA, not a PHP regex translation. @internal */
final readonly class PatternProgram
{
    /** @param list<PatternState> $states */
    public function __construct(public int $start, public array $states) {}
}

/** @internal */
final readonly class PatternState
{
    /** @param list<array{int, int}> $ranges */
    public function __construct(
        public string $op,
        public int $target = -1,
        public int $first = -1,
        public int $second = -1,
        public array $ranges = [],
    ) {}
}

/** Validate a selected schema root against exact JSON; no defaults or coercions. */
final class Validator
{
    public static function validate(int $root, JsonValue $value): void { (new ValidationSession())->check($root, $value); }
}

/** Shared budgets survive all branch trials and cannot become ordinary invalidity. @internal */
final class ValidationSession
{
    use ScopedValidation;
    use ResourceValidation;
    private readonly bool $scoped;
    /** @var list<ValidationNode> */
    private readonly array $nodes;
    /** @var array<int, true> */
    private readonly array $roots;
    /** @var array<string, true> */
    private array $active = [];
    private int $steps = RuntimeConfig::MAX_EVALUATION_STEPS;
    private int $equalities = RuntimeConfig::MAX_EQUALITY_STEPS;
    private int $numeric = RuntimeConfig::MAX_EVALUATION_STEPS;

    public function __construct(private readonly ?CallControl $control = null)
    {
        $this->scoped=$this->profile(ValidationProgram::VERSION, ValidationProgram::PROFILE);
        $this->resources=$this->resourceMetadata(ValidationProgram::VERSION,ValidationProgram::class);
        $this->nodes = ValidationProgram::nodes();
        $this->roots = array_fill_keys(ValidationProgram::roots(), true);
    }
    private function profile(string $version, string $profile): bool
    {
        if($version==='suspect.validation.experimental.v1'&&$profile==='oas31-jsonschema202012-static-subset'){return false;}
        if($version==='suspect.validation.experimental.v2'&&$profile==='oas31-jsonschema202012-static-applicators'){return true;}
        if($version==='suspect.validation.experimental.v3'&&$profile==='oas31-jsonschema202012-resources-dynamic'){return true;}
        $this->failure('', '', 'unknown validation program version/profile');
    }

    public function check(int $root, JsonValue $value): void
    {
        if (!isset($this->roots[$root])) { $this->failure('', '', 'schema is not a selected validation root'); }
        $issue = $this->scoped?$this->evaluateScoped($root,$value,'',0)->issue:$this->evaluate($root, $value, '', 0);
        if ($issue !== null) { throw $issue; }
    }

    /** Internal native union selection shares this session's finite budgets. */
    public function matches(int $root, JsonValue $value): bool
    {
        if (!isset($this->roots[$root])) { $this->failure('', '', 'schema is not a selected validation root'); }
        return ($this->scoped?$this->evaluateScoped($root,$value,'',0)->issue:$this->evaluate($root, $value, '', 0)) === null;
    }

    private function failure(string $source, string $path, string $message): never
    {
        throw new ValidationError('evaluation_failure', $source, $path, $message);
    }
    private function spend(string $source, string $path, int $amount = 1): void
    {
        $this->control?->check();
        $this->steps -= $amount;
        if ($this->steps < 0) { $this->failure($source, $path, 'evaluation work budget exhausted'); }
    }
    private function child(string $path, string $key): string { return $path . '/' . str_replace(['~', '/'], ['~0', '~1'], $key); }

    private function evaluate(int $index, JsonValue $value, string $path, int $depth): ?ValidationError
    {
        if (!isset($this->nodes[$index])) { $this->failure('', $path, 'unknown instruction target'); }
        $node = $this->nodes[$index]; $this->spend($node->source, $path);
        $identity = $index . ':' . spl_object_id($value);
        if ($depth >= RuntimeConfig::MAX_SCHEMA_DEPTH || isset($this->active[$identity])) {
            $this->failure($node->source, $path, 'evaluation depth/nonproductive recursion limit');
        }
        $this->active[$identity] = true;
        try {
            $first = null;
            foreach ($node->checks as $check) {
                $this->spend($check->source, $path);
                // Do not short circuit: a later evaluation failure dominates
                // even an already-invalid conjunct or already-valid union.
                $issue = $this->instruction($check, $value, $path, $depth);
                $first ??= $issue;
            }
            return $first;
        } finally { unset($this->active[$identity]); }
    }

    private function mismatch(ValidationInstruction $check, string $path, string $message = ''): ValidationError
    {
        return new ValidationError('invalid', $check->source, $path, $message === '' ? $check->op . ' assertion rejected the value' : $message);
    }
    private function number(JsonNumber $number, ValidationInstruction $check, string $path): JsonNumber
    {
        if (strlen($number->token) > RuntimeConfig::MAX_NUMBER_BYTES) { $this->failure($check->source, $path, 'numeric operand byte budget exhausted'); }
        $this->numeric -= strlen($number->token);
        if ($this->numeric < 0) { $this->failure($check->source, $path, 'numeric work budget exhausted'); }
        return $number;
    }
    private function operand(ValidationInstruction $check, string $path): JsonNumber
    {
        if ($check->number === null) { $this->failure($check->source, $path, 'missing numeric operand'); }
        return $this->number($check->number, $check, $path);
    }

    private function literal(ValidationLiteral $literal, ValidationInstruction $check, string $path): JsonValue
    {
        try { return $literal->value($this->control); }
        catch (JsonError $e) { $this->failure($check->source, $path, 'literal representation exceeds the native JSON profile'); }
    }

    private function instruction(ValidationInstruction $check, JsonValue $value, string $path, int $depth): ?ValidationError
    {
        $kind = $value->kind; $result = true;
        switch ($check->op) {
            case 'always': $result = $check->truth; break;
            case 'type':
                $result = in_array($kind->value, $check->types, true);
                if (!$result && $kind === JsonKind::Number && in_array('integer', $check->types, true)) {
                    $result = $this->number($value->asNumber(), $check, $path)->isInteger();
                }
                break;
            case 'ref': return $this->evaluate($check->target, $value, $path, $depth + 1);
            case 'properties':
            case 'additionalProperties':
                if ($kind !== JsonKind::Object) { break; }
                $members = $value->asObject(); $first = null;
                if ($check->op === 'properties') {
                    foreach ($check->properties as [$name, $target]) {
                        $this->spend($check->source, $path);
                        if (array_key_exists($name, $members)) {
                            $issue = $this->evaluate($target, $members[$name], $this->child($path, $name), $depth + 1);
                            $first ??= $issue;
                        }
                    }
                } else {
                    $declared = array_fill_keys($check->declared, true);
                    foreach ($members as $name => $item) {
                        $this->spend($check->source, $path);
                        if (!isset($declared[$name])) {
                            $issue = $this->evaluate($check->target, $item, $this->child($path, (string) $name), $depth + 1);
                            $first ??= $issue;
                        }
                    }
                }
                return $first;
            case 'required':
                if ($kind === JsonKind::Object) {
                    $members = $value->asObject(); $first = null;
                    foreach ($check->names as $name) {
                        $this->spend($check->source, $path);
                        if (!array_key_exists($name, $members)) { $first ??= $this->mismatch($check, $path, 'required member is absent: ' . $name); }
                    }
                    return $first;
                }
                break;
            case 'items':
            case 'prefixItems':
                if ($kind === JsonKind::Array) {
                    $first = null;
                    foreach ($value->asArray() as $index => $item) {
                        if ($check->op === 'items' && $index < $check->start) { continue; }
                        if ($check->op === 'prefixItems' && !isset($check->targets[$index])) { break; }
                        $this->spend($check->source, $path);
                        $target = $check->op === 'items' ? $check->target : $check->targets[$index];
                        $issue = $this->evaluate($target, $item, $this->child($path, (string) $index), $depth + 1);
                        $first ??= $issue;
                    }
                    return $first;
                }
                break;
            case 'allOf':
            case 'anyOf':
            case 'oneOf':
                $matches = 0;
                foreach ($check->targets as $target) {
                    $this->spend($check->source, $path);
                    if ($this->evaluate($target, $value, $path, $depth + 1) === null) { ++$matches; }
                }
                $result = match ($check->op) { 'allOf' => $matches === count($check->targets), 'anyOf' => $matches > 0, default => $matches === 1 };
                break;
            case 'not': $result = $this->evaluate($check->target, $value, $path, $depth + 1) !== null; break;
            case 'bound':
                if ($kind === JsonKind::Number) {
                    $order = $this->number($value->asNumber(), $check, $path)->compare($this->operand($check, $path));
                    $result = ($check->maximum ? $order <= 0 : $order >= 0) && (!$check->exclusive || $order !== 0);
                }
                break;
            case 'multipleOf':
                if ($kind === JsonKind::Number) {
                    $result = $this->number($value->asNumber(), $check, $path)->multipleOf($this->operand($check, $path), function (int $cost) use ($check, $path): void {
                        $this->numeric -= $cost;
                        if ($this->numeric < 0) { $this->failure($check->source, $path, 'numeric work budget exhausted'); }
                    });
                }
                break;
            case 'count':
                if ($kind->value === $check->countTarget) {
                    $count = match ($kind) {
                        JsonKind::Array => count($value->asArray()), JsonKind::Object => count($value->asObject()),
                        JsonKind::String => self::scalarCount($value->asString()), default => 0,
                    };
                    $order = JsonNumber::fromInt($count)->compare($this->operand($check, $path));
                    $result = $check->maximum ? $order <= 0 : $order >= 0;
                }
                break;
            case 'const':
                if ($check->constant === null) { $this->failure($check->source, $path, 'missing constant operand'); }
                $result = $this->equal($value, $this->literal($check->constant, $check, $path), $check, $path, 0); break;
            case 'enum':
                $result = false;
                foreach ($check->literals as $literal) {
                    $this->spend($check->source, $path);
                    if ($this->equal($value, $this->literal($literal, $check, $path), $check, $path, 0)) { $result = true; break; }
                }
                break;
            case 'uniqueItems':
                if ($kind === JsonKind::Array) {
                    $items = $value->asArray();
                    for ($i = 0, $length = count($items); $i < $length; ++$i) {
                        $this->spend($check->source, $path);
                        for ($j = 0; $j < $i; ++$j) {
                            $this->spend($check->source, $path);
                            if ($this->equal($items[$i], $items[$j], $check, $path, 0)) { return $this->mismatch($check, $path); }
                        }
                    }
                }
                break;
            case 'pattern':
                if ($kind === JsonKind::String) {
                    if ($check->pattern === null) { $this->failure($check->source, $path, 'missing pattern program'); }
                    $result = $this->pattern($check->pattern, $value->asString(), $check, $path);
                }
                break;
            default: $this->failure($check->source, $path, 'unknown instruction: ' . $check->op);
        }
        return $result ? null : $this->mismatch($check, $path);
    }

    private function equal(JsonValue $a, JsonValue $b, ValidationInstruction $check, string $path, int $depth): bool
    {
        $this->control?->check();
        if (--$this->equalities < 0 || $depth > RuntimeConfig::MAX_SCHEMA_DEPTH) { $this->failure($check->source, $path, 'equality budget exhausted'); }
        if ($a->kind !== $b->kind) { return false; }
        switch ($a->kind) {
            case JsonKind::Null: return true;
            case JsonKind::Boolean: return $a->asBool() === $b->asBool();
            case JsonKind::String: return $a->asString() === $b->asString();
            case JsonKind::Number: return $this->number($a->asNumber(), $check, $path)->compare($this->number($b->asNumber(), $check, $path)) === 0;
            case JsonKind::Array:
                $left = $a->asArray(); $right = $b->asArray();
                if (count($left) !== count($right)) { return false; }
                foreach ($left as $i => $item) { if (!$this->equal($item, $right[$i], $check, $path, $depth + 1)) { return false; } }
                return true;
            case JsonKind::Object:
                $left = $a->asObject(); $right = $b->asObject();
                if (count($left) !== count($right)) { return false; }
                foreach ($left as $key => $item) {
                    if (!array_key_exists($key, $right) || !$this->equal($item, $right[$key], $check, $path, $depth + 1)) { return false; }
                }
                return true;
        }
    }

    /** JSON construction has already verified Unicode scalar encodings. */
    private static function scalarCount(string $text): int
    {
        $count = 0;
        for ($i = 0, $length = strlen($text); $i < $length; ++$i) { if ((ord($text[$i]) & 0xC0) !== 0x80) { ++$count; } }
        return $count;
    }

    private function pattern(PatternProgram $program, string $text, ValidationInstruction $check, string $path): bool
    {
        $this->spend($check->source, $path, strlen($text));
        $chars = preg_split('//u', $text, -1, PREG_SPLIT_NO_EMPTY);
        if ($chars === false) { $this->failure($check->source, $path, 'invalid Unicode instance'); }
        $length = count($chars);
        for ($offset = 0; $offset <= $length; ++$offset) {
            $current = [$program->start => true];
            for ($position = $offset; $position <= $length; ++$position) {
                $pending = array_keys($current); $seen = []; $consuming = [];
                while ($pending !== []) {
                    $this->spend($check->source, $path);
                    $index = array_pop($pending);
                    if (isset($seen[$index])) { continue; } $seen[$index] = true;
                    if (!isset($program->states[$index])) { $this->failure($check->source, $path, 'unknown pattern target'); }
                    $state = $program->states[$index];
                    switch ($state->op) {
                        case 'match': return true;
                        case 'split': $pending[] = $state->first; $pending[] = $state->second; break;
                        case 'jump': $pending[] = $state->target; break;
                        case 'start': if ($position === 0) { $pending[] = $state->target; } break;
                        case 'end': if ($position === $length) { $pending[] = $state->target; } break;
                        case 'char': $consuming[] = $state; break;
                        default: $this->failure($check->source, $path, 'unknown pattern instruction');
                    }
                }
                if ($position === $length) { break; }
                $point = self::codepoint($chars[$position]); $current = [];
                foreach ($consuming as $state) {
                    foreach ($state->ranges as [$low, $high]) {
                        $this->spend($check->source, $path);
                        if ($point >= $low && $point <= $high) { $current[$state->target] = true; break; }
                    }
                }
                if ($current === []) { break; }
            }
        }
        return false;
    }

    private static function codepoint(string $char): int
    {
        $head = ord($char[0]);
        return match (strlen($char)) {
            1 => $head,
            2 => (($head & 31) << 6) | (ord($char[1]) & 63),
            3 => (($head & 15) << 12) | ((ord($char[1]) & 63) << 6) | (ord($char[2]) & 63),
            default => (($head & 7) << 18) | ((ord($char[1]) & 63) << 12) | ((ord($char[2]) & 63) << 6) | (ord($char[3]) & 63),
        };
    }
}

File: crates/suspect-codegen/src/cpp_sdk/validation.cpp
#include "@PACKAGE@/runtime.hpp"
#include "number.hpp"

namespace @NAMESPACE@::detail {
static CodecError invalid(const Source& source, const std::string& path, std::string message) {
    return {CodecError::Kind::Validation, source, path, std::move(message), std::nullopt};
}
static std::string_view kind(const JsonValue& value) {
    switch (value.value.index()) {
        case 0: return "null"; case 1: return "boolean"; case 2: return "number";
        case 3: return "string"; case 4: return "array"; case 5: return "object";
        default: fail(CodecError::Kind::EvaluationFailure, {}, "", "valueless JSON variant");
    }
}
Validation::Validation(Control control) : Validation(validation_program(), control) {}
Validation::Validation(const Program& checked_program, Control control) : program_(checked_program), control_(control), steps_(program_.max_evaluation_steps), equality_(program_.max_equality_steps), numeric_(program_.max_evaluation_steps) {
    resources_=program_.version=="suspect.validation.experimental.v3"&&program_.profile=="oas31-jsonschema202012-resources-dynamic";
    scoped_=resources_||(program_.version=="suspect.validation.experimental.v2"&&program_.profile=="oas31-jsonschema202012-static-applicators");
    if(!scoped_&&(program_.version!="suspect.validation.experimental.v1"||program_.profile!="oas31-jsonschema202012-static-subset"))
        fail(CodecError::Kind::EvaluationFailure,{},"","unknown validation version/profile pair");
    if(!scoped_)for(const auto& node:program_.nodes)for(const auto& check:node.checks)if(check.op>=Op::If)
        fail(CodecError::Kind::EvaluationFailure,check.source,"","v2 opcode in a v1 program");
    check_resources();
}
void Validation::spend(std::size_t& budget, std::size_t cost, const Source& source, const std::string& path) {
    control_.check(source, path);
    if (cost > budget) fail(CodecError::Kind::EvaluationFailure, source, path, "validation evaluation/equality/numeric work exhausted");
    budget -= cost;
}
bool Validation::matches(std::size_t root, const JsonValue& value, const std::string& path) {
    if (!program_.roots.contains(root)) fail(CodecError::Kind::EvaluationFailure, {}, path, "schema root is not selected");
    numeric_checked_.clear();
    return scoped_?!evaluate_v2(root,value,path,0).failure:!evaluate(root, value, path, 0);
}
void Validation::require(std::size_t root, const JsonValue& value, const std::string& path) {
    if (!program_.roots.contains(root)) fail(CodecError::Kind::EvaluationFailure, {}, path, "schema root is not selected");
    numeric_checked_.clear();
    auto failure = scoped_?evaluate_v2(root,value,path,0).failure:evaluate(root, value, path, 0);
    if (failure) throw Failure{std::move(*failure)};
}
std::optional<CodecError> Validation::evaluate(std::size_t index, const JsonValue& value, const std::string& path, std::size_t depth) {
    if (index >= program_.nodes.size()) fail(CodecError::Kind::EvaluationFailure, {}, path, "unknown schema target");
    auto& node = program_.nodes[index];
    spend(steps_, 1, node.source, path);
    if (depth >= program_.max_depth) fail(CodecError::Kind::EvaluationFailure, node.source, path, "schema depth limit");
    auto key = std::pair{index, &value};
    if (!active_.insert(key).second) fail(CodecError::Kind::EvaluationFailure, node.source, path, "recursive schema revisited the same instance without progress");
    struct ActiveGuard { decltype(active_)& active; decltype(key) key_; ~ActiveGuard() { active.erase(key_); } } guard{active_, key};
    std::optional<CodecError> first;
    for (auto& check : node.checks) {
        spend(steps_, 1, check.source, path);
        auto outcome = instruction(check, value, path, depth);
        if (!first && outcome) first = std::move(outcome);
    }
    return first;
}
std::optional<CodecError> Validation::instruction(const Check& check, const JsonValue& value, const std::string& path, std::size_t depth) {
    auto bad = [&]() { return invalid(check.source, path, "schema assertion rejected value"); };
    auto number = [&](std::string_view token) {
        if (token.size() > program_.max_number_bytes) fail(CodecError::Kind::EvaluationFailure, check.source, path, "validation exact-number operand byte limit");
        spend(numeric_, token.size(), check.source, path);
        return Decimal::parse(token);
    };
    auto conjunction = [&](const auto& each) -> std::optional<CodecError> {
        std::optional<CodecError> first;
        each([&](std::size_t target, const JsonValue& child, const std::string& at) {
            spend(steps_, 1, check.source, path);
            auto outcome = evaluate(target, child, at, depth + 1);
            if (!first && outcome) first = std::move(outcome);
        });
        return first;
    };
    switch (check.op) {
        case Op::Always: if (!check.flag) return bad(); break;
        case Op::Type: {
            for (auto& allowed : check.names) {
                if (allowed == kind(value)) return std::nullopt;
                if (allowed == "integer" && value.is<JsonNumber>() && number(value.as<JsonNumber>().token()).integral()) return std::nullopt;
            }
            return bad();
        }
        case Op::Ref: return evaluate(check.target, value, path, depth + 1);
        case Op::Properties:
            if (value.is<JsonValue::Object>()) return conjunction([&](auto child) {
                for (auto& property : check.properties) {
                    spend(steps_, 1, check.source, path);
                    auto it = value.as<JsonValue::Object>().find(property.name);
                    if (it != value.as<JsonValue::Object>().end()) child(property.target, it->second, child_path(path, property.name));
                }
            });
            break;
        case Op::AdditionalProperties:
            if (value.is<JsonValue::Object>()) return conjunction([&](auto child) {
                for (auto& [key, v] : value.as<JsonValue::Object>()) {
                    spend(steps_, 1, check.source, path);
                    if (std::find(check.names.begin(), check.names.end(), key) == check.names.end()) child(check.target, v, child_path(path, key));
                }
            });
            break;
        case Op::Required:
            if (value.is<JsonValue::Object>()) {
                for (auto& key : check.names) {
                    spend(steps_, 1, check.source, path);
                    if (!value.as<JsonValue::Object>().contains(key)) return invalid(check.source, child_path(path, key), "required member absent");
                }
            }
            break;
        case Op::Items:
            if (value.is<JsonValue::Array>()) return conjunction([&](auto child) {
                auto& array = value.as<JsonValue::Array>();
                for (std::size_t i = check.start; i < array.size(); ++i) child(check.target, array[i], child_path(path, std::to_string(i)));
            });
            break;
        case Op::PrefixItems:
            if (value.is<JsonValue::Array>()) return conjunction([&](auto child) {
                auto& array = value.as<JsonValue::Array>();
                for (std::size_t i = 0; i < std::min(array.size(), check.targets.size()); ++i) child(check.targets[i], array[i], child_path(path, std::to_string(i)));
            });
            break;
        case Op::AllOf: return conjunction([&](auto child) { for (auto target : check.targets) child(target, value, path); });
        case Op::AnyOf: case Op::OneOf: {
            std::size_t matches = 0;
            for (auto target : check.targets) {
                spend(steps_, 1, check.source, path);
                if (!evaluate(target, value, path, depth + 1)) ++matches;
            }
            if (matches == 0 || (check.op == Op::OneOf && matches != 1)) return bad();
            break;
        }
        case Op::Not: if (!evaluate(check.target, value, path, depth + 1)) return bad(); break;
        case Op::Bound: case Op::MultipleOf:
            if (value.is<JsonNumber>()) {
                auto a = number(value.as<JsonNumber>().token()); auto b = number(check.operand);
                if (check.op == Op::Bound) {
                    int order = a.compare(b);
                    if ((check.flag && order > 0) || (!check.flag && order < 0) || (check.exclusive && order == 0)) return bad();
                } else if (!a.divisible(b, [&](std::size_t cost) { spend(numeric_, cost, check.source, path); })) return bad();
            }
            break;
        case Op::Count:
            if (kind(value) == check.operand) {
                auto size = value.is<std::string>() ? unicode_length(value.as<std::string>()) :
                    value.is<JsonValue::Array>() ? value.as<JsonValue::Array>().size() : value.as<JsonValue::Object>().size();
                // Cardinalities are compiled symbolic counts, independent of
                // the numeric-instance operand-byte policy.
                auto a = Decimal::parse(std::to_string(size)); auto b = Decimal::parse(check.names.at(0));
                int order = a.compare(b);
                if ((check.flag && order > 0) || (!check.flag && order < 0)) return bad();
            }
            break;
        case Op::Const:
            if (!equal(value, check.literals.at(0), check.source, path, 0)) return bad();
            break;
        case Op::Enum:
            for (auto& literal : check.literals) {
                spend(steps_, 1, check.source, path);
                if (equal(value, literal, check.source, path, 0)) return std::nullopt;
            }
            return bad();
        case Op::UniqueItems:
            if (value.is<JsonValue::Array>()) {
                auto& array = value.as<JsonValue::Array>();
                for (std::size_t i = 0; i < array.size(); ++i) {
                    spend(steps_, 1, check.source, path);
                    for (std::size_t j = 0; j < i; ++j) {
                        spend(steps_, 1, check.source, path);
                        if (equal(array[i], array[j], check.source, path, 0)) return bad();
                    }
                }
            }
            break;
        case Op::Pattern:
            if (value.is<std::string>() && !pattern(check.pattern, value.as<std::string>(), check.source, path)) return bad();
            break;
        default: fail(CodecError::Kind::EvaluationFailure, check.source, path, "unknown validation instruction");
    }
    return std::nullopt;
}
bool Validation::equal(const JsonValue& a, const JsonValue& b, const Source& source, const std::string& path, std::size_t depth) {
    spend(equality_, 1, source, path);
    if (depth > program_.max_depth) fail(CodecError::Kind::EvaluationFailure, source, path, "equality nesting limit");
    if (kind(a) != kind(b)) return false;
    if (a.is<Null>()) return true;
    if (a.is<bool>()) return a.as<bool>() == b.as<bool>();
    if (a.is<std::string>()) return a.as<std::string>() == b.as<std::string>();
    if (a.is<JsonNumber>()) {
        if (a.as<JsonNumber>().token().size() > program_.max_number_bytes || b.as<JsonNumber>().token().size() > program_.max_number_bytes)
            fail(CodecError::Kind::EvaluationFailure, source, path, "equality exact-number operand byte limit");
        spend(numeric_, a.as<JsonNumber>().token().size() + b.as<JsonNumber>().token().size(), source, path);
        return a.as<JsonNumber>() == b.as<JsonNumber>();
    }
    if (a.is<JsonValue::Array>()) {
        auto& left = a.as<JsonValue::Array>(); auto& right = b.as<JsonValue::Array>();
        if (left.size() != right.size()) return false;
        for (std::size_t i = 0; i < left.size(); ++i) if (!equal(left[i], right[i], source, path, depth + 1)) return false;
        return true;
    }
    auto& left = a.as<JsonValue::Object>(); auto& right = b.as<JsonValue::Object>();
    if (left.size() != right.size()) return false;
    for (auto& [key, v] : left) {
        auto it = right.find(key);
        if (it == right.end() || !equal(v, it->second, source, path, depth + 1)) return false;
    }
    return true;
}

bool Validation::pattern(const PatternProgram& program, std::string_view text, const Source& source, const std::string& path) {
    spend(steps_, 1, source, path);
    // Streaming Unicode scalar iteration avoids an input-sized scalar array.
    std::vector<std::size_t> seeds;
    std::size_t offset = 0;
    for (;;) {
        std::vector<std::size_t> pending = std::move(seeds);
        pending.push_back(program.start);
        std::set<std::size_t> seen;
        std::vector<std::size_t> consuming;
        while (!pending.empty()) {
            spend(steps_, 1, source, path);
            auto index = pending.back(); pending.pop_back();
            if (index >= program.states.size()) fail(CodecError::Kind::EvaluationFailure, source, path, "invalid pattern state");
            if (!seen.insert(index).second) continue;
            auto& state = program.states[index];
            switch (state.kind) {
                case PatternState::Kind::Match: return true;
                case PatternState::Kind::Split: pending.push_back(state.second); pending.push_back(state.first); break;
                case PatternState::Kind::Jump: pending.push_back(state.first); break;
                case PatternState::Kind::Start: if (offset == 0) pending.push_back(state.first); break;
                case PatternState::Kind::End: if (offset == text.size()) pending.push_back(state.first); break;
                case PatternState::Kind::Char: consuming.push_back(index); break;
            }
        }
        if (offset == text.size()) return false;
        spend(steps_, 1, source, path);
        auto point = unicode_scalar(text, offset);
        seeds.clear();
        for (auto index : consuming) {
            auto& state = program.states[index];
            for (auto [low, high] : state.ranges) {
                spend(steps_, 1, source, path);
                if (point < low) break;
                if (point <= high) { seeds.push_back(state.first); break; }
            }
        }
    }
}
} // namespace @NAMESPACE@::detail
